### PR TITLE
introduce universes to NLL type check

### LINKED ISSUE #48071: ICE bootstrapping rustc with NLL enabled
When I try to bootstrap rustc with NLL enabled, I get an odd ICE compiling `libsyntax_ext`:

```
error: internal compiler error: broken MIR in NodeId(16117) (_368 = move _369): bad assignment (for<'cx, 'r, 's> fn(&'cx mut syntax::ext::base::ExtCtxt<'r>, syntax_pos::Span, &'s [syntax::tokenstream::To\
kenTree]) -> std::boxed::Box<syntax::ext::base::MacResult + 'cx> = for<'r, 's, 't0> fn(&'r mut syntax::ext::base::ExtCtxt<'s>, syntax_pos::Span, &'t0 [syntax::tokenstream::TokenTree]) -> std::boxed::Box<\
syntax::ext::base::MacResult>): RegionsOverlyPolymorphic(BrNamed(crate11:DefIndex(1:1012), 'cx), '_#729r)
   --> src/libsyntax_ext/lib.rs:65:44
    |
65  |                           expander: Box::new($f as MacroExpanderFn),
    |                                              ^^^^^^^^^^^^^^^^^^^^^
...
92  | /     register! {
93  | |         line: expand_line,
94  | |         __rust_unstable_column: expand_column_gated,
95  | |         column: expand_column,
...   |
112 | |         compile_error: compile_error::expand_compile_error,
113 | |     }
    | |_____- in this macro invocation
```

### PR DESCRIPTION
**Pull request by nikomatsakis**
This branch aims to fix #48071 and also advance chalk integration a bit at the same time. It re-implements the subtyping/type-equating check so that NLL doesn't "piggy back" on the subtyping code of the old type checker. 

This new code uses the "universe-based" approach to handling higher-ranked lifetimes, which sidesteps some of the limitations of the current "leak-based" scheme. This avoids the ICE in #48071.

At the same time, I aim for this to potentially be a kind of optimization. This NLL code is (currently) not cached, but it also generates constraints without doing as much instantiation, substitution, and folding. Right now, though, it still piggy backs on the `relate_tys` trait, which is a bit unfortunate -- it means we are doing more hashing and things than we have to. I want to measure the see the perf. Refactoring that trait is something I'd prefer to leave for follow-up work.

r? @pnkfelix -- but I want to measure perf etc first

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Ah, shoot, the conflicts here are a bit "deep-ish". Have to think about best way to resolve.

**Comment by nikomatsakis** [comment]
@bors try

still, I think we can do a try run to measure perf...

**Comment by kennytm** [comment]
@rust-timer build b5045d96d659e3aa4c727f44caf5362cc6116be3

**Comment by kennytm** [comment]
Perf is ready. Tuple-stress is extremely slowed down (+50%), and also html5ever (+7%). OTOH coercions (-10%) and clap (-8%) are improved a lot.

**Comment by nikomatsakis** [comment]
Interesting. I'll have to check out tuple-stress.

**Comment by nikomatsakis** [comment]
Curious. I cannot reproduce this. Although I think my build is from before the rebase. Let me try again.

File: src/Cargo.lock [hunk 1]
```diff
@@ -87,7 +87,7 @@ dependencies = [
 
 [[package]]
 name = "atty"
-version = "0.2.11"
+version = "0.2.10"
```
Comment (pnkfelix) [inline src/Cargo.lock]: weird, why did this happen?

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs [hunk 2]
```diff
@@ -50,90 +51,65 @@ impl RegionValueElements {
...
-    /// Converts a particular `RegionElementIndex` to a universal
-    /// region, if that is what it represents. Returns `None`
-    /// otherwise.
-    crate fn to_universal_region(&self, i: RegionElementIndex) -> Option<RegionVid> {
-        if i.index() < self.num_universal_regions {
-            Some(RegionVid::new(i.index()))
-        } else {
-            None
+    /// Converts a particular `RegionElementIndex` to a location, if
+    /// that is what it represents. Returns `None` otherwise.
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/nll/region_infer/values.rs]: The sentence "Returns `None` otherwise." Can't be right since you've made this return `Location`, not an `Option<_>`...

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs [hunk 3]
```diff
@@ -202,73 +153,110 @@ impl<N: Idx> RegionValues<N> {
...
     /// True if `sup_region` contains all the CFG points that
     /// `sub_region` contains. Ignores universal regions.
-    crate fn contains_points(&self, sup_region: N, sub_region: N) -> bool {
+    crate fn contains_points(
+        &self,
+        sup_region: N,
+        sub_region: N,
+    ) -> bool {
         // This could be done faster by comparing the bitsets. But I
         // am lazy.
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/nll/region_infer/values.rs]: Regarding the comment above the code below *is* now comparing bitsets, right?

File: src/test/ui/lifetime-errors/ex3-both-anon-regions-2.nll.stderr [hunk 4]
```diff
@@ -5,12 +5,12 @@ LL |     *v = x; //~ ERROR lifetime mismatch
...
 error[E0623]: lifetime mismatch
-  --> $DIR/ex3-both-anon-regions-2.rs:11:14
+  --> $DIR/ex3-both-anon-regions-2.rs:12:5
    |
 LL | fn foo(&mut (ref mut v, w): &mut (&u8, &u8), x: &u8) {
-   |              ^^^^^^^^^            ---           --- these two types are declared with different lifetimes...
-   |              |
-   |              ...but data from `x` flows here
+   |                                   ---           --- these two types are declared with different lifetimes...
+LL |     *v = x; //~ ERROR lifetime mismatch
```
Comment (pnkfelix) [inline src/test/ui/lifetime-errors/ex3-both-anon-regions-2.nll.stderr]: woo hoo!

File: src/test/ui/lifetime-errors/ex2e-push-inference-variable-3.nll.stderr [hunk 5]
```diff
@@ -5,12 +5,13 @@ LL |     let b = Ref { data: y.data };
...
-  --> $DIR/ex2e-push-inference-variable-3.rs:16:9
+  --> $DIR/ex2e-push-inference-variable-3.rs:18:5
    |
 LL | fn foo<'a, 'b, 'c>(x: &'a mut Vec<Ref<'b, i32>>, y: Ref<'c, i32>) {
    |                                   ------------      ------------ these two types are declared with different lifetimes...
-LL |     let a: &mut Vec<Ref<i32>> = x; //~ ERROR lifetime mismatch
-   |         ^ ...but data from `y` flows into `x` here
+...
+LL |     Vec::push(a, b);
+   |     ^^^^^^^^^^^^^^^ ...but data from `y` flows into `x` here
```
Comment (pnkfelix) [inline src/test/ui/lifetime-errors/ex2e-push-inference-variable-3.nll.stderr]: double woo hoo!

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs [hunk 6]
```diff
@@ -98,6 +97,12 @@ impl RegionValueElements {
 /// graph. Constructed efficiently from `RegionValueElements`.
 newtype_index!(PointIndex { DEBUG_FORMAT = "PointIndex({})" });
 
+/// A single integer representing a (non-zero) `UniverseIndex`.
+/// Computed just by subtracting one from `UniverseIndex`; this is
+/// because the `0` value for `UniverseIndex` represents the root
+/// universe, and we don't need/want a bit for that one.
+newtype_index!(PlaceholderIndex { DEBUG_FORMAT = "PointIndex({})" });
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/nll/region_infer/values.rs]: You should change the `DEBUG_FORMAT` here to `PlaceholderIndex({})`.

File: src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+            self.relate(a.skip_binder(), b.skip_binder())?;
+
+            self.b_scopes.pop().unwrap();
+            self.a_scopes.pop().unwrap();
+        }
+
+        if self.ambient_contravariance() {
+            // Contravariance, so we want `for<'a> A :> for<'b...> B`
+            // -- do the opposite, therefore, of covariant. Replace
+            // the `'a...` placeholders with `'b...` variables.
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs]: I find the phrasing "replace the `'a...` placeholders with `'b...` variables" confusing.

Would it be acceptable to instead say "check if any instantiation of B (where `'b...` have been replaced with existentials) can match against any instantiation of A (where `'a...` have been replaced with placeholders)", which more closely matches the wording you used above (which I understand better)?

Or is there some reason you chose this particular phrasing here?

File: src/test/ui/nll/relate_tys/hr-fn-aau-eq-abu.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test an interesting corner case: a fn that takes two arguments that
+// are references with the same lifetime is in fact equivalent to a fn
+// that takes two references with distinct lifetimes. This is true
```
Comment (pnkfelix) [inline src/test/ui/nll/relate_tys/hr-fn-aau-eq-abu.rs]: ... you should probably revise this comment, or at least elaborate a bit that while in principle such fn's are equivalent, our checks are currently distinguishing them?

Comment (pnkfelix) [reply src/test/ui/nll/relate_tys/hr-fn-aau-eq-abu.rs]: (or are you figuring that the FIXME down in the body suffices for that purpose? I just find the difference between the header vs the actual behavior jarring.)

File: src/test/ui/nll/relate_tys/hr-fn-aba-as-aaa.rs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test that the NLL `relate_tys` code correctly deduces that a
+// function returning always its first argument can be upcast to one
+// that returns either first or second argument.
```
Comment (pnkfelix) [inline src/test/ui/nll/relate_tys/hr-fn-aba-as-aaa.rs]: oh wow this is actually kind of exciting now that I see it.

File: src/Cargo.lock (hunk 1 above)
Comment (nikomatsakis) [reply src/Cargo.lock]: Uh..I have no idea. I should revert that.

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/region_infer/values.rs]: Yeah, outdated. Will fix.

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs (hunk 3 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/region_infer/values.rs]: Er yes should fix :)

File: src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs (hunk 7 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs]: Your way seems clearer, will fix.

File: src/test/ui/nll/relate_tys/hr-fn-aau-eq-abu.rs (hunk 8 above)
Comment (nikomatsakis) [reply src/test/ui/nll/relate_tys/hr-fn-aau-eq-abu.rs]: Yeah, will fix.

**Comment by pnkfelix** [comment]
r=me once nits addressed and travis passes.

**Comment by nikomatsakis** [comment]
I'm marking this as S-waiting-on-perf because I remember it had some perf impact that I was unable to reproduce earlier, and I want to check about that. The PR has changed significantly in the meantime so the earlier results may no longer apply.

File: src/librustc_mir/borrow_check/nll/region_infer/values.rs [hunk 10]
```diff
@@ -142,193 +111,328 @@ crate enum RegionElement {
...
+/// `LivenessValues`, which only stores those points in the where a
+/// region is live). The full value for a region may contain points in
+/// the CFG, but also free regions as well as bound universe
+/// placeholders.
+///
+/// Example:
+///
+/// ```rust
+/// fn foo(x: &'a u32) -> &'a u32 {
+///    let y: &'0 u32 = x; // let's call this `'0`
```
Comment (kennytm) [inline src/librustc_mir/borrow_check/nll/region_infer/values.rs]:
```
[00:56:06] error: character literal may only contain one codepoint: '0
[00:56:06]  --> <stdin>:2:12
[00:56:06]   |
[00:56:06] 2 |    let y: &'0 u32 = x; // let's call this `'0`
[00:56:06]   |            ^^
```

Make this ` ```text ` or something.

**Comment by kennytm** [comment]
@rust-timer build 73f817a0577649c3ffd1045b12c45f406a41a47f

**Comment by nikomatsakis** [comment]
OK so perf looks great. Now just gotta fix the last trailing errors.

**Comment by nikomatsakis** [comment]
@bors r=pnkfelix p=1

Resolves a major ICE for NLL, EP2 blocker.

File: src/Cargo.lock (hunk 1 above)
Comment (RalfJung) [reply src/Cargo.lock]: But you did not... the downgrade landed.

Comment (RalfJung) [reply src/Cargo.lock]: Submitted fix as https://github.com/rust-lang/rust/pull/52779
